### IDs:
Insert yours IDs to the cell below

ID #1:

ID #2:


## Read the following instructions carefully:

1. This jupyter notebook contains all the step by step instructions needed for this exercise.
1. You are free to add cells.
1. Write your functions and your answers in this jupyter notebook only.
1. Answers to theoretical questions should be written in **markdown cells (with $\LaTeX$ support)**.
1. Submit this jupyter notebook only using your ID as a filename. Not to use ZIP or RAR. For example, your Moodle submission file name should look like this (two id numbers): `123456789_987654321.ipynb`.

### Question 1 - Defective products

In a manufacturing pipeline products are 3% defective. We are interested in examining a defective product to see what goes wrong on the belt. We need to ask the facility manager to send us a set of independent samples for examination.

#### 1.A

How many independent samples should we ask for in order to have a 85% probability of having at least one defective product in the batch sent? You should write a function.

#### 1.B
Answer this part again with the following changes: products are 4% defective and we want a 95% probability of at least one defective product in the batch.

#### 1.C 

Consider the following cases and calculate how many independent samples are required: 

1. Products are 10% defective and we want a 90% probability of at least 5 defective products in the batch.
1. Products are 30% defective and we want a 90% probability of at least 15 defective products in the batch.

Explain the difference between the two results. You should use mathematical reasoning based on the properties of distributions you saw in class and visualizations in your answer.

### Question 2 - Rent distributions in Randomistan

The state of Randomistan conducted a survey to study the distribution of rent paid in two neighboring towns, Stochastic Heights and Random Grove, to be denoted SH and RG.<br> 

Here are some findings of the survey:
* The population of SH and RG is 16,000 and 22,000 respectively. <br>
* The mean rent in SH and RG is 6300RCU and 4200RCU respectively.
* The median rent is 4600RCU in both towns.
* The IQR of the rent is smaller in SH than in RG.

All data generated in this question needs to be consistent with these findings.

#### 2.A
Draw histograms that describe 2 different scenarii of possible distributions of rent in the two towns.
Your histograms should:<br>
* Use bins of 100RCU each.
* Have at least 10 non zero bins.

#### 2.B
Draw a histogram of a third scenario with the same properties. <br>
In addition, in this scenario the rent in SH should have a higher variance than the rent in RG.

The survey also examined the per household income (PHI) in these two places.<br>

It found that:<br>
* The mean of PHI in SH is 12500 and in RG is 8500.
* The median is 12000 in SH and 8000 in RG.
* The covariance of the rent and the PHI was observed to be as in the formula below with $\alpha=97\%$ and $\alpha=89\%$ in SH and in RG respectively.<br><br>
$$Cov(rent, PHI) = \alpha * \sqrt{Var(rent)} * \sqrt{Var(PHI)}$$

#### 2.C
Produce rent and PHI data for the two cities, that is consistent with these findings. The covariances in your data can deviate by up to 1% from the numbers given $\alpha$.

#### 2.D
Produce two heatmaps that describe these two bivariate joint distributions. Make sure you carefully consider the selected binning resolution.

### Question 3 - Multinomial Distributions

1. Let $X \sim Multinomial(n,\vec{p})$ be a multinomial random variable where $n=20$ and $\vec{p} = (0.2,  0.1,  0.1,  0.1,  0.2,  0.3)$. Note that X is a vector of counts.


2. Let $Y = X_2 + X_3 + X_4$ be a random variable.


3. Create $k=100$ experiments where $X$ is sampled using Python. Calculate the empirical centralized third moment of $Y$ based on your $k$ experiments.


4. Compare your result to the calculation in class for the centralized third moment of the **binomial** distribution and explain your observation.

In [72]:
import numpy as np 

np.random.seed(666)

X = np.random.multinomial(size = 100, n = 20, pvals=[0.2,  0.1,  0.1,  0.1,  0.2,  0.3])
Y = X[:,[1,2,3]].sum(axis=1)



In [73]:
emp_cent_mom_3 = (np.power((Y - Y.mean()), 3)).sum() / Y.shape[0]
emp_cent_mom_3


1.6785695171917845

$Y = X_2 + X_3 + X_4 \sim B(20, p_2 + p_3 + p_4) = B(20, 0.3)$

The formula seen in class states that the empirical centralized third moment of the binomial distribution is $ np(1-p)(1-2p)$

In [70]:
n = 20 
p = 0.3 
binom_n_p_emp_cent_mom_3 = n*p*(1-p)*(1-2*p)
binom_n_p_emp_cent_mom_3

1.6799999999999997

In [76]:
print("Distance between the third moment that we expect from the formula and from a randomly generated Y: " +
      f"{abs(emp_cent_mom_3 - binom_n_p_emp_cent_mom_3)}")

Distance between the third moment that we expect from the formula and from a randomly generated Y: 0.0014304828082152277


Notice that we have chosen a seed which we chose to be 666 and happens to leads to the Y vector which yields a third moment of 1.679. This is very close to the theoretical 1.68. However it is not true that every seed will produce a third moment that generates a value close to the theoretical one on k = 100. On the other hand, as we increase k, the experiment automatically converges to the theoretical value 1.68

### Question 4 - Covariance and independence

What is the variance of the sum X +Y + Z of three random variables in terms of the variances of X, Y and Z and the covariances between each pair of random variables? What happens if X,Y,Z are pairwise independent? If X,Y,Z are pairwise independent, are they necessarily collectively independent? Prove your answer.

$V(X+Y+Z) = V((X+Y)+Z) = V(X+Y) + V(Z) + 2Cov(X+Y, Z)$

 By bilinearity: 
 $$ = V(X) + V(Y) + V(Z) + 2Cov(X, Z) + 2Cov(Y, Z) + 2Cov(X, Y) \ \ \ \ (2)$$

If X, Y and Z are pairwise independent then each of the covariance pair (except $Cov(D,D)$ where $D \in \{X,Y,Z\}$) is 0. This implies that $V(X+Y+Z) =  V(X) + V(Y) + V(Z)$ by (2).

Let $X, Y \sim B(1,0.5)$ and $Z = X \oplus Y$

$P(X=x) = P(Y=y) = P(Z=z) = \dfrac{1}{2}, P(X=x, Y=y) = P(X=x, Y=y) =  P(X=x, Y=y) = \dfrac{1}{4}$ 

$= \dfrac{1}{2} \cdot \dfrac{1}{2} =  P(D_1=d_1)P(D_2=d_2) $ where $D_1, D_2 \in \{X,Y,Z\}$ clearly showing parwise independence.

Now consider $P(X=0,Y=0,Z=1) = 0$ as $ 0 \oplus 0 = 0 \neq 1$ but $P(X=0)P(Y=0)P(Z=1)=\dfrac{1}{2} \cdot \dfrac{1}{2} \cdot \dfrac{1}{2} = \dfrac{1}{8} \neq 0$ proving that they are not collectively dependent $\square$.




### Question 5 - Convolutions

#### 5.A
Write a program, `Q = NFoldConv(P , n)`, that takes as input:
* A distribution, P, of a random variable that takes finitely many integer values
* An integer n

and produces the distribution, Q, of the sum of n independent repeats of random variables, each of which has the distribution P.

In [1]:
import numpy as np
import pandas as pd
from itertools import product
import matplotlib.pyplot as plt


def NFoldConv(P , n):
    t = {}
    t['values'] = np.array(list(product(P['values'], repeat=n))).sum(axis=1)
    t['prob'] = np.prod(list(product(P['prob'], repeat=n)),axis=1)
    return pd.DataFrame(t).groupby(by='values').sum()

#### 5.B
Compute the distribution of the sum of the results of rolling a fair octahedron 17 times.

<img src="https://upload.wikimedia.org/wikipedia/commons/2/27/Octahedron.jpg" width="200">


In [ ]:
P={}
P['values'] = np.array([1,2,3,4,5,6,7,8])
P['prob'] = np.array([1/8 for i in range(8)])
n=17

df = NFoldConv(P, n)

df.plot.bar(figsize=(20,10),title='distribution of the sum of the results of rolling a fair octahedron 17 times')